File: kattis/2016 ICPC Mid Atlantic/H Painting the Floodwall/paint.py
dp = {}
seen = []

def maxBefore(x: int) -> int:
    lo, hi = 0, len(seen) - 1
    if hi == -1:
        return 0
    while hi - lo > 1:
        mid = (hi + lo) // 2
        if seen[mid] <= x:
            lo = mid
        else:
            hi = mid
    if seen[hi] <= x:
        return seen[hi]
    return seen[lo]

while True:
    try:
        n = int(input())
    except:
        exit(0)
    ranges = []
    for _ in range(n):
        x, y = map(int, input().split())
        ranges.append((y, x))
    ranges = sorted(ranges)
    dp = {0: 0}
    seen = [0]
    for y, x in ranges:
        ans = dp[maxBefore(y)]
        ans = max(ans, y - x + dp[maxBefore(x)])
        dp[y] = ans
        seen.append(y)
    print(dp[seen[-1]])

File: kattis/2016 ICPC Mid Atlantic/H Painting the Floodwall/paint.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef pair<ll, ll> pll;
map<ll, ll> dp;
vector<ll> seen;

ll maxBefore (ll x) {
    auto tmp = upper_bound(seen.begin(), seen.end(), x);
    tmp --;
    return *tmp;
}

int main() {
    int n;
    while (cin >> n) {
        vector<pll> ranges;
        while (n --) {
            ll x, y;
            cin >> x >> y;
            ranges.push_back({y, x});
        }
        sort(ranges.begin(), ranges.end());
        dp.clear();
        dp[0] = 0;
        seen.clear();
        seen.push_back(0);
        ll y, x;
        for (auto p: ranges) {
            y = p.first; x = p.second;
            ll ans = dp[maxBefore(y)];
            ans = max(ans, y - x + dp[maxBefore(x)]);
            dp[y] = ans;
            seen.push_back(y);
        }
        // for (auto p: dp) {
        //     cerr << p.first << " " << p.second << endl;
        // }

        cout << dp[y] << endl;
    }
}
